File: attentive-android-sdk/src/test/java/com/attentive/androidsdk/events/CustomEventTest.java
package com.attentive.androidsdk.events;

import static org.junit.Assert.assertThrows;

import java.util.Arrays;
import java.util.Collection;
import java.util.Map;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.Parameterized;

@RunWith(Parameterized.class)
public class CustomEventTest {
    private final String type;
    private final Map<String, String> properties;

    public CustomEventTest(String type, Map<String, String> properties) {
        this.type = type;
        this.properties = properties;
    }

    @Test
    public void customEventBuilder() {
        System.out.printf("'%s', '%s'%n", type, properties);
        new CustomEvent.Builder(type, properties);
    }

    @Parameterized.Parameters
    public static Collection<Object[]> provideCustomEventBuildParams() {
        return Arrays.asList(new Object[][] {
            { null, Map.of() },
            { "", Map.of() },
            { "f", null },
            { "typeWithInvalidChar[", Map.of() },
            { "f", Map.of("keyWithInvalidChar[", "value") },
            { "f", Map.of() },
            { "f", Map.of("key", "value") }
        });
    }
}

File: attentive-android-sdk/src/test/java/com/attentive/androidsdk/events/CustomEventTest.kt
package com.attentive.androidsdk.events

import org.junit.Assert
import org.junit.Test
import org.junit.runner.RunWith
import org.junit.runners.Parameterized

@RunWith(Parameterized::class)
class CustomEventTest(
    val type: String,
    val properties: Map<String, String>,
    val throwableClass: Class<Throwable>?,
) {
    @Test
    fun customEventBuilder() {
        System.out.printf("'%s', '%s', '%s'%n", type, properties, throwableClass)
        if (throwableClass == null) {
            // no throw
            CustomEvent.Builder(type, properties)
        } else {
            Assert.assertThrows(
                throwableClass,
            ) {
                CustomEvent.Builder(
                    type,
                    properties,
                ).build()
            }
        }
    }

    companion object {
        @Parameterized.Parameters
        @JvmStatic
        fun provideCustomEventBuildParams(): Collection<Array<Any?>> {
            return listOf(
                arrayOf("", emptyMap<String, String>(), IllegalArgumentException::class.java),
                arrayOf("typeWithInvalidChar[", emptyMap<String, String>(), IllegalArgumentException::class.java),
                arrayOf("f", mapOf("keyWithInvalidChar[" to "value"), IllegalArgumentException::class.java),
                arrayOf("f", emptyMap<String, String>(), null),
                arrayOf("f", mapOf("key" to "value"), null),
            )
        }
    }
}
